<?xml version="1.0"?>
<launch>
  <arg name="log_type" default="screen"/>

  <!-- Caster description -->
  <param name="robot_description" command="$(find xacro)/xacro --inorder $(find caster_description)/urdf/caster.urdf.xacro" />

  <!-- Socketcan bridge -->
  <node pkg="socketcan_bridge" type="socketcan_bridge_node" name="socketcan_bridge_node" output="$(arg log_type)">
    <param name="can_device" type="string" value="can0" />
  </node>

  <!-- Caster driver -->
  <node pkg="caster_base" type="caster_base_socketcan_node" name="caster_base_node" output="$(arg log_type)">
    <param name="can_send_topic" type="string" value="sent_messages" />
    <param name="can_receive_topic" type="string" value="received_messages" />
    <param name="can_id" type="int" value="1" />
    <param name="left_wheel_joint" type="string" value="drive_wheel_left_joint" />
    <param name="right_wheel_joint" type="string" value="drive_wheel_right_joint" />
    <remap from="caster_velocity_controller/cmd_vel" to="cmd_vel"/>
    <remap from="caster_velocity_controller/odom" to="odom"/>
  </node>

  <!-- Imu driver -->
  <node pkg="jy901_driver" type="jy901_driver_node" name="jy901_node" output="$(arg log_type)" >
    <param name="baudrate" type="int" value="115200" />
    <param name="port" type="string" value="/dev/jy901" />
    <param name="looprate" type="int" value="40" />
    <remap from="imu/data" to="imu_data"/>
  </node>

  <!-- Laser driver -->
  <node pkg="urg_node" type="urg_node" name="urg_node" output="$(arg log_type)">
    <param name="ip_address" value="192.168.33.7"/>
    <param name="frame_id" value="laser_link"/>
    <param name="calibrate_time" value="false"/>
    <param name="publish_intensity" value="true"/>
    <param name="publish_multiecho" value="false"/>
    <param name="angle_min" value="-1.4835299"/>
    <param name="angle_max" value="1.4835299"/>
  </node>

  <!-- rplidar driver -->
  <node name="rplidarNode"          pkg="rplidar_ros"  type="rplidarNode" output="screen">
    <param name="serial_port"         type="string" value="/dev/rplidar"/>
    <param name="serial_baudrate"     type="int"    value="115200"/>
    <param name="frame_id"            type="string" value="laser_link2"/>
    <param name="inverted"            type="bool"   value="false"/>
    <param name="angle_compensate"    type="bool"   value="true"/>
    <remap from="scan" to="scan_2"/>
  </node>

  <!-- BMS driver -->
  <node pkg="hongfu_bms_status" type="hongfu_bms_status_node" name="hongfu_bms_status_node" output="$(arg log_type)">
    <param name="port" type="string" value="/dev/caster_bms" />
    <param name="frame_id" type="string" value= "hongfu_bms" />
    <param name="baudrate" type="int" value="9600" />
    <param name="looprate" type="int" value="1" />
    <remap from="/hongfu_bms_status/hongfu_bms" to="/bms"/>
  </node>

  <!-- Ultrasonic sensor driver -->
  <node pkg="dauxi_ks106" type="dauxi_ks106_node" name="dauxi_ks106_node" output="screen">
    <param name="port" type="string" value="/dev/caster_us"/>
    <param name="baudrate" type="int" value="9600"/>
    <param name="frequency" type="int" value="10"/>
    <param name="min_range" type="double" value="0.20" />
    <param name="max_range" type="double" value="1.50" />
    <param name="topic" type="string" value="us"/>
    <param name="us1_id" type="string" value="ultrasonic_front_left"/>
    <param name="us2_id" type="string" value="ultrasonic_front_right"/>
    <param name="us3_id" type="string" value="ultrasonic_rear_right"/>
    <param name="us4_id" type="string" value="ultrasonic_rear_left"/>
  </node>

  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="$(arg log_type)">
    <param name="publish_frequency" type="double" value="30" />
  </node>

  <!-- Caster controller -->
  <include file="$(find caster_control)/launch/control.launch">
    <arg name="log_type" value="$(arg log_type)"/>
  </include>
  
  <!-- Robot pose ekf -->
<!--   <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf" output="$(arg log_type)">
    <param name="output_frame" value="odom"/>
    <param name="base_footprint_frame" value="base_footprint"/>
    <param name="freq" value="50.0"/>
    <param name="sensor_timeout" value="1.0"/>
    <param name="odom_used" value="true"/>
    <param name="imu_used" value="true"/>
    <param name="vo_used" value="false"/>
    <param name="debug" value="true"/>
    <param name="self_diagnose" value="false"/>
  </node> -->

  <!--Teleop -->
  <!-- <include file="$(find caster_control)/launch/teleop.launch" /> -->

  <!-- Viz -->
  <!-- <node name="rviz" pkg="rviz" type="rviz" args="-d $(find caster_viz)/viz/view.rviz" /> -->
</launch>